server/game/core/GameObject.ts

```diff
@@ -25,6 +25,12 @@ export abstract class GameObject {
     }
 
     public addOngoingEffect(ongoingEffect: IOngoingCardEffect) {
+        // logic for "lose keyword" abilites: they need to register all effects that are giving the keyword at the time they are applied, so that they can nullify only those
+        if (ongoingEffect.type === EffectName.LoseKeyword) {
```

Comment:
This works but it's not the best practice to add complexity in base classes like this which is specific for certain situations with certain subclasses. Separating the implementation from the context can make it harder to avoid bugs, and also if folks are working on the code they may not realize stuff is happening at this layer (or they may not know where to find this implementation if they're looking for it).

Typically when an ongoing effect has some complex logic like this we tend to create a class implementation for handling the complexity (e.g. GainAbility or CostAdjuster), I think there's enough here to justify creating a small one.

The other benefit of that would be offering maybe slightly better documentation and typing for the properties that we're stuffing into the effect value, since that system is untyped it can be a little hard to understand what the contents of an effect are in each situation so having it clearly documented is a big win.

At the barest minimum though this logic should be moved into an override of `addOngoingEffect` in UnitProperties since that's the only class it applies to afaik

Reply:
Thank you. I figured there was a more isolated way to define effects like this but between OngoingEffect itself, IOngoingCardEffect, OngoingEffectImpl, and the ValueWrapper(which I don't think I even noticed before) I kind of had no idea what was what.

Reply:
Yeah that's totally fair, the ongoing effect system is one of the most tangled parts of the codebase. They cheated _a lot_ with types so it's going to be a pretty big project to unravel at all at some point. The general idea is that OngoingEffect holds an OngoingEffectImpl, which in turn holds a ValueWrapper. IOngoingCardEffect is an interface that partially matches OngoingEffect that they use to hide some of the bad stuff they did with types.

Reply:
I made a LoseKeyword class extending OngoingEffectValueWrapper(like GainAbility, which you mentioned), but it looks like I won't be able to access anything in it in getKeywords() because in that context all the ongoingEffects are IOngoingCardEffect, which doesn't know anything about ValueWrappers. I don't want to mess with the underlying stuff to change that because I'm still new to this system and don't really understand it. Does that make sense and do you have any idea what I should do?

Reply:
Oh yeah so I was thinking that it would operate very similar to the way GainAbility does in which the ongoing effect acts on the unit instead of the unit reading from the effect. So e.g. for GainAbility it tells the unit to add a new ability to itself, with the blanking it would tell the unit to apply the blanking effect to all of its current keywords using something like the algorithm I described in the comment on Card.ts (though there are several options there).

If you did want to continue your current route and instead have the card read from the effect, then you should take a look at the stat modification system where we use the modifyStats() effect to wrap a StatsModifier object which then gets unwrapped in UnitProperties.getPower() / getHp().

Reply:
Okay. I didn't think of just adding a method on Card, good idea.

Reply:
I ended up not needing to add a method actually, but the idea of the LoseKeyword asking Card for information rather than the other way around was helpful.